### PR TITLE
Add S.M.A.PowerShell.Create method overload with Runspace argument

### PR DESCRIPTION
**Pull request by KirkMunro**
## PR Summary

When invoking multiple PowerShell commands one at a time from an assembly that does not have PowerShell loaded by default, developers should create a runspace first via `RunspaceFactory.Create` inside of a using statement, and then for each command they want to invoke, create a PowerShell instance via `PowerShelll.Create` from within another using statement, and then after that set the runspace for the PowerShell instance to the runspace they created from the RunspaceFactory. This is cumbersome and unintuitive, and really should be simplified to make it easier to use and understand.

This PR adds a new `PowerShell.Create` method overload that accepts a Runspace object so that it is easier for developers to follow this model in their code.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/3131
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
@KirkMunro It would be great if you would link two gists with samples before and after the change to demo benefits.

**Comment by adityapatwardhan** [comment]
@KirkMunro I believe this can be done using the overload for `IntialSessionState`. Which in effect creates a runspace and opens it. Am I missing something?

https://github.com/PowerShell/PowerShell/blob/4e5e3900f646449e91a30434dc4f5576329d82b2/src/System.Management.Automation/engine/hostifaces/PowerShell.cs#L853-L861

**Comment by KirkMunro** [comment]
@adityapatwardhan That only works for one invocation. A single PowerShell instance. The PR allows you to create your runspace and initialize it with the initialSessionState, and then easily use it over, and over, for as many PowerShell instances as you need, keeping each instance in the same runspace that is initialized only once.

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Should we consider checking runspace state? Also, to match the other overloads behavior should the runspace be opened is it is in 'BeforeOpening' state?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (KirkMunro) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Runspace state checks are already built-in when you invoke PowerShell.Invoke. Plus, there is no need to check state here. You can create the PowerShell instance regardless of the runspace state.

For the second question, this overload of the Create method takes a runspace that you have created and are managing elsewhere. In the other overload that you are referring to (the one in InitialSessionState), that overload creates the runspace that it assigns to the PowerShell instance, and since it is being done from the InitialSessionState that you want to load, it makes sense to open the runspace because that is the only way to load that session state into the runspace. In this case, opening the runspace is a task that is done elsewhere so it would be wrong to call open again.

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: I think this will be very useful.  However, we should always validate parameter arguments on public methods.
```
if (runspace == null)
{
    throw new System.Management.Automation.PSArgumentNullException("runspace");
}
```
I agree that the provided runspace does not need to be in an "Opened" state, however, if it is opened then we should check its Availability.  It is useless if the Availability is "Busy" or "RemoteDebug".  If it is "AvailableForNestedCommand" then the IsNested property must be true, but you can run a nested command only if the PowerShell object is running on the same thread.  You can reuse existing error messages.

Also, if RunspaceState is "Disconnected", then this is useless ... there is a separate method for getting a PowerShell object that can reconnect to a remote session.  Similar issue for "Broken" or "Closed" states.  I think it would be good to provide the user an error immediately when the runspace is clearly not in a valid state for assignment to the PowerShell object.  "BeforeOpen" state is Ok because that can be corrected.  But the other states, there is nothing the use can do and I feel an error should be immediately thrown.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Seems other create methods don't check runspace state.

We should use `nameof` in `PSArgumentNullException(nameof(runspace))`.

Comment (PaulHigin) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: "Seems other create methods don't check runspace state."

I guess it is a matter of opinion.  You can create a PowerShell object without any runspace association at all, but it seems to me that a static method that specifically constructs a PowerShell object for a provided runspace, should ensure that the returned object is viable.  I think of it as input parameter validation, if the runspace is not usable then it is not a valid input to the method.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee is ok with this change

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Code can't predict the future.  It can only work with the data on hand.  It seems wrong to me to successfully return an object known to be invalid.  The user then spends effort composing a possibly complex serious of statements only to have it inevitably fail.  As a user I prefer to know that the returned object is invalid up front.

But I am Ok with just checking for null exception on the runspace parameter argument.

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: I see your point. My thoughts was that we create the runspace object _before_ Powershell object and can/should check a state of the runspace before creating Powershell object. I'd expect such user behavior.
 I am Ok with just checking for null exception too.

Comment (KirkMunro) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Thanks for this great discussion guys. I really like the null check because the intent here is very clearly to use an existing runspace, so having it check against null makes great sense. I added a null check and a Pester test to ensure it throws when passed null.

**Review by PaulHigin — APPROVED** [review]
LGTM

File: test/powershell/engine/Api/BasicEngine.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/engine/Api/BasicEngine.Tests.ps1]:
```suggestion
            { [powershell]::Create($null) } | Should -Throw -ErrorId 'PSArgumentNullException'
```

**Comment by KirkMunro** [comment]
@SteveL-MSFT What's the hold up on this one? Is it simply outstanding code reviews?

**Comment by SteveL-MSFT** [comment]
@KirkMunro Yes, please address the code review feedback.  Otherwise, it's good to go.

**Comment by KirkMunro** [comment]
@SteveL-MSFT Unless I'm missing something there are no outstanding code review changes to make. See the discussion between Paul, Ilya and I.

**Comment by SteveL-MSFT** [comment]
@KirkMunro you are correct.  Unfortunately we're a bit understaffed right now due to the holidays.  I expect this to be merged after the new year.

@adityapatwardhan should review your changes and approve

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: The comment should be more specific to this overload, like `Construct an empty PowerShell instance and associate it with the given Runspace`.

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: I think it makes sense to not check for Runspace state and perform actions like calling `Open` when it's `BeforeOpening`, because it's possible the Runspace is being opened in another thread while we are creating a `PowerShell` instance using this method. Let's just be dumb and behave more predictably.

However, I think adding a comment here to explain why we don't check for state or so would be very useful.

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: nitpick: this new line can be removed.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (KirkMunro) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Done.

File: src/System.Management.Automation/engine/hostifaces/PowerShell.cs
Comment (KirkMunro) [reply src/System.Management.Automation/engine/hostifaces/PowerShell.cs]: Agreed, both done. Thanks for the review!

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@daxian-dbw Could you update your review please?

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by adityapatwardhan** [comment]
@daxian-dbw I merged the change. I thought it was assigned to me.

**Pull request merged**
